Field report: CardCoins rbf and batching (#582)
bitcoinops/bitcoinops.github.io: _includes/articles/cardcoins-rbf-batching.md

{:.post-meta}
*by [CardCoins][]*

> 

Reviewer: If we want to include an short definition at the opening of the blog post, here is a suggestion: 

"Additive batching" is a scheme whereby additional outputs are included to unconfirmed transactions in the mempool. This field report outlines efforts CardCoins has in implementing a reorg and DoS safe implementation of such a scheme.

Will let you guys ultimately decide on its inclusion. I could go either way.

> [Replace By Fee][topic rbf] (RBF, BIP-125) and [batching][payment batching] are two important tools for any enterprises

Reviewer 2: Nit: use our normal BIP styling; the consistency is especially helpful to me when I'm `git grep`ing for things we previously wrote.

```suggestion
[Replace By Fee][topic rbf] (RBF, BIP125) and [batching][payment batching] are two important tools for any enterprises
```

Reviewer 2: Also needs edit later in the article (not important for git grep, but important for intra-article consistency).

directly interacting with Bitcoin's mempool. Fees go up, fees go down, but
the business must always fight for fee efficiency.

Each tool, while powerful, has its own complexities and nuances. For example,
batching customer withdrawals may save on fees for the enterprise, but will likely make
[child pays for parent][topic cpfp] (CPFP) uneconomical for a customer who wishes to speed up the transaction.
Similarly, RBF is useful for an enterprise who takes a fee-underbidding strategy
(their initial transaction broadcast starts at a low fee, and is slowly bid
upwards), but it exposes their customers to [potential confusion][rbf blog] as their
withdrawal transaction updates in their wallet. It would also be messy for the
customer to spend from this transaction while it remains unconfirmed, as the

> enterprise will have to account for this child spend when attempting to replace

Reviewer 2: Nit: I think this is more specific about the immediate burden placed on the enterprise.

```suggestion
enterprise will have to pay for this child spend when attempting to replace
```

the parent. Even worse, the enterprise may have a withdrawal [pinned][pinning] by another

> service that the customer withdrew their funds directly to.

Reviewer 2: Nitpick.

```suggestion
service which received the customer's withdrawal.
```

When combining these two tools, a service provider unlocks new functionality but
is similarly exposed to novel forms of complexity. In the base case, combining
RBF and a single, static batch carries a simple combination of the complexities
that RBF and batching carry discretely. However, when you combine RBF and
"additive batching," emergent edge cases and dangerous failure scenarios present
themselves.

Additive RBF batching is a transaction construction protocol that leverages the
rules of BIP-125 to allow a transactor to introduce new outputs (and confirmed

> inputs) to a transaction in the mempool. This enables an enterprise to give

Reviewer 3: Maybe this definition or something similar could be put in the opening paragraph to introduce the subject.

the mempool. This transaction continues to be updated until it confirms or
reaches some other local optimum.

There are many strategies to this type of additive RBF batching. At [CardCoins][] we
took a safety-first approach to our implementation (with the help of [Matthew

> Zipkin][]), the nuances of which we detailed in a more detailed blog post, [RBF

Reviewer: Minor edit, remove the extra 'detailed': "the nuances of which we detailed in the blog post,"
